crates/miden-lib/asm/miden/contracts/auth/multisig.masm

```diff
@@ -1,24 +1,208 @@
 
 # Event to place the falcon signature of a provided message and public key on the advice stack.
-const.FALCON_SIG_TO_STACK=131087
+const.AUTH_REQUEST=131087
+const.UNAUTHORIZED_EVENT=131100
+
+const.THRESHOLD_SLOT=0
+const.NUMBER_OF_APPROVERS_SLOT=1
+const.PUBLIC_KEYS_MAP_SLOT=2
+const.EXECUTED_TXS_SLOT=3
+
+# CONSTANTS
+# AdviceMap Keys
+
+# Runtime Memory Addresses
+const.LOOP_ITER=100
+const.TEMP_PK=108
+const.MESSAGE=112
+const.THRESHOLD_VALUE=116
+const.OWNER_PUB_KEY=120
+
+# ERRORS
+const.NON_CONSTANT_MSG="NON CONSTANT SIGNATURES"
+const.INVALID_SIGNATURE="SIG DOES NOT MATCH DIGEST"
+
+# => []
+
+proc.verify_signatures_loop
+    push.0
+    # => [initial_iteration_value]
+
+    mem_store.LOOP_ITER
+    # => []
+
+    push.1
+    while.true
+        mem_load.LOOP_ITER
+        # => [iteration]
+
+        push.0.0.0
+        # => [KEY]
+
+        push.PUBLIC_KEYS_MAP_SLOT
+        # => [index, KEY]
+
+        exec.account::get_map_item
+        # => [OWNER_PUB_KEY]
+
+        mem_storew.TEMP_PK
+        # => [OWNER_PUB_KEY]
+
+        padw mem_loadw.MESSAGE
+        # => [MESSAGE, OWNER_PUB_KEY]
+
+        hmerge
+        # => [SIG_KEY]
+
+        adv.has_mapkey
+        # => [SIG_KEY]
+
+        adv_push.1 not
+        # => [unauthorized, SIG_KEY]
+
+        # If signatures are non-existant the tx will fail here:
+        if.true
+            mem_loadw.MESSAGE padw
+            # => [EMPTY_WORK, MSG]
+
+            emit.UNAUTHORIZED_EVENT
+        end
+        # => [SIG_KEY]
+
+        adv.push_mapval
+        # => [SIG_KEY]
+
+        dropw
+        # => []
+
+        mem_loadw.MESSAGE padw mem_loadw.TEMP_PK
+        # => [PK, MSG]
+
+        # @dev not sure if this is correct (could be authenticator signature?)...
+        emit.AUTH_REQUEST
+
+        exec.rpo_falcon512::verify
+        # => []
+
+        mem_load.LOOP_ITER
+        # => [iteration]
+
+        mem_load.THRESHOLD_VALUE
+
+        lt
+        if.true
+            mem_load.LOOP_ITER
+            push.1 add
+            mem_store.LOOP_ITER
+
+            push.1
+        else
+            push.0
+        end
+    end
+
+end
 
-# The slot in this component's storage layout where the public key is stored.
-const.PUBLIC_KEY_SLOT=0
+# => [MSG]
 
-    
-#! Authenticate a transaction using the Falcon signature scheme.
-#! This is a single auth procedure that handles both the addition of a signature and the multisig tx execution.
-#! 
+proc.replay_protection.4
+    dupw loc_storew.0 dropw
+
+    push.EXECUTED_TXS_SLOT
+    # => [index, MSG]
+
+    push.111 debug.stack drop
+
+    exec.account::get_map_item
+    # => [LOOKUP_VAL]
+
+    padw
+    # => [EMPTY_WORD, LOOKUP_VAL]
+
+    assert_eqw
+    # => []
+
+    push.1.1.1.1
+
+    loc_loadw.0
+
+    push.EXECUTED_TXS_SLOT
+    # => [index, KEY, VALUE]
+
+    # Setting the key value pair in the map
+    exec.account::set_map_item
+
+    dropw dropw dropw dropw
+end
```

Comment:
what's the purpose of `replay_protection`? The `SALT` word should take care of this

Reply:
ah yes, thanks 👍🏼